Add options to disable telemetry modules

ankitkumarr opened the pull request:
Helps with #2123 

I still need to test this out in a private site extensions.

The issue talks about allowing selection of perf counters to use. For now, I didn't make that change. As we should discuss if we want to allow users to have a black-list or a white-list. And counter names are like `"\\Process(??APP_WIN32_PROC??)\\% Processor Time Normalized"`. Not sure if we want users to have strings like that in the `host.json` to configure that. They seem too specific to do a string mapping.

@brettsam, thoughts?

ankitkumarr committed:
Add options to disable telemetry modules

cijothomas commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -38,6 +38,18 @@ public class ApplicationInsightsLoggerOptions : IOptionsFormatter
+        /// Gets or sets the flag that enables quick pulse collection.
+        /// Enabled by default.
+        /// </summary>
+        public bool EnableQuickPulse { get; set; } = true;
```
Suggest to use the name LiveMetrics as thats the name of public feature. (QuickPulse was an internal name)
https://docs.microsoft.com/en-us/azure/azure-monitor/app/live-stream

ankitkumarr committed:
Register QuickPulseTelemetry before disabling

ankitkumarr commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -105,7 +117,9 @@ public string Format()
                 { nameof(EnablePerformanceCountersCollection), EnablePerformanceCountersCollection},
                 { nameof(HttpAutoCollectionOptions), httpOptions},
-                { nameof(QuickPulseInitializationDelay), QuickPulseInitializationDelay }
+                { nameof(QuickPulseInitializationDelay), QuickPulseInitializationDelay },
```
I need to update this name as well

ankitkumarr committed:
Update QuickPulse external string to LiveMetrics

brettsam commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -36,7 +36,7 @@ public class ApplicationInsightsLoggerOptions : IOptionsFormatter
         /// can result in decreased application startup time. Default value is 15 seconds.
         /// </summary>
-        public TimeSpan QuickPulseInitializationDelay { get; set; } = TimeSpan.FromSeconds(15);
+        public TimeSpan LiveMetricsInitializationDelay { get; set; } = TimeSpan.FromSeconds(15);
```
Thanks :-)

brettsam replied:
Shoot -- I don't think we can change this as this has been released already. So -- can you create a second property called `LiveMetricsInitializationDelay` -- and mark the `QuickPulse` one as deprecated? We just won't ever use `QuickPulse` one internally.

brettsam replied:
... it was never documented but it makes me nervous that someone could build locally against one version and deploy against another with a missing property.

ankitkumarr replied:
Ohh I didn't realize it had been released. Fair, I will make that update. Working through a couple of release issue. May get delayed until Monday

ankitkumarr replied:
Looks like there's also this setting `QuickPulseAuthenticationApiKey` that i need to update
But that's been there for a while. We probably need to keep supporting it as well.

brettsam commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/ApplicationInsightsServiceCollectionExtensions.cs:
```diff
@@ -257,9 +271,9 @@ internal static LoggerFilterOptions CreateFilterOptions(LoggerFilterOptions regi
+                    delayer.ScheduleInitialization(() => module.Initialize(configuration), options.LiveMetricsInitializationDelay);
                 }
-                else
+                else if (module != null)
```
Can the module ever be null here? You can get the NullTelmetryModule.Instance, but that should just be a no-op.

ankitkumarr replied:
Yes, at least in our tests they can be null. One of the instance in `telemetryModules` was null. I think it's because the Host setup test explicitly removes an `ITelemetryModule`. I figured it's a harmless check to have.

brettsam reviewed (approved):
This looks good to me! 

FYI for @lmolkova and @cijothomas -- we're adding properties to allow people to turn a couple of modules off (via host.json in functions). Let us know if you have any objections...

brettsam commented:
Let's get these two new ones in, then we can come back to chat about perf counters. Feel free to split off the perf counters into a separate issue for tracking purposes going forward (make sure to note it in the existing issue).

cijothomas commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -45,6 +45,18 @@ public class ApplicationInsightsLoggerOptions : IOptionsFormatter
         public bool EnablePerformanceCountersCollection { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets the flag that enables quick pulse collection.
```
minor: comment also should say Live Metrics

ankitkumarr replied:
Oh right, my bad

cijothomas reviewed (approved):
Please modify the doc to advice customers of the new ability to turn off speific features.
https://docs.microsoft.com/en-us/azure/azure-functions/functions-monitoring#disable-built-in-logging

ankitkumarr committed:
Deprecate QuickPulse to LiveMetrics

ankitkumarr commented:
@brettsam, I updated the PR and added `[Deprecated]` in the comments. I didn't find any attribute or something that we use to deprecate flags. So figured, this would be ok. Lmk what you think

brettsam commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -28,23 +28,48 @@ public class ApplicationInsightsLoggerOptions : IOptionsFormatter
         public SnapshotCollectorConfiguration SnapshotConfiguration { get; set; }
 
         /// <summary>
+        /// [Deprecated] Please use LiveMetricsAuthenticationApiKey.
```
Use `[Obsolete]` here -- it lets you put in a string for a description and just say "Use ... instead."

ankitkumarr replied:
Awesome, that's what I was looking for. Thanks :)

brettsam commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLoggerOptions.cs:
```diff
@@ -28,23 +28,48 @@ public class ApplicationInsightsLoggerOptions : IOptionsFormatter
         /// Gets or sets the time to delay Quick Pulse (Live Metrics) initialization. Delaying this initialization
         /// can result in decreased application startup time. Default value is 15 seconds.
         /// </summary>
         public TimeSpan QuickPulseInitializationDelay { get; set; } = TimeSpan.FromSeconds(15);
```
Let's make this getter and setter just use LiveMetricsInitializationDelay directly -- so there's only one source of truth until we delete it.

Something like: 
`get { return LiveMetricsInitializationDelay; } set { LiveMetricsInitializationDelay = value; }`

Same goes for the auth api key...

brettsam reviewed (approved):
couple small comments...
